[4.x] Fix user groups/roles querying

ryanmitchell opened the pull request:
This PR updates the state store paths for user groups to use `->` rather than `/` which resolves the errors in #6124 and follows up on the comments in #2498.

It also updates the `getKeysWithWhere` method in the `UserQueryBuilder` to have logic for roles and groups so that we do a comparison on their handle, rather than the entire object (which was failing). Not sure how this fits in with Eloquent... I might need to update this logic or handle it a different way?

Fixes: #6124.
Fixes: #2498.
Fixes: #7578

jasonvarga commented on src/Auth/File/UserGroup.php:
```diff
@@ -9,6 +9,6 @@ class UserGroup extends BaseUserGroup
     public function queryUsers()
     {
-        return User::query()->where('groups/'.$this->handle(), true);
+        return User::query()->where('groups->'.$this->handle(), true);
```
I would rather not do this as it would be a breaking change.

Anyone doing `->where('groups/foo')` would need to change to `->where('groups->foo')`.

I think a better solution would be to escape the slashes when we go to add the items in the cache.

In this line, we could (also) replace slashes with something else: https://github.com/statamic/cms/blob/f6b573ee5046b55f75bb450c69fe2ec5529eca34/src/Stache/Indexes/Index.php#L131

ryanmitchell replied:
Ok I've opted for `->` to keep it consistent with whats been done with JSON fields

jasonvarga commented on src/Stache/Indexes/Users/Group.php:
```diff
@@ -9,7 +9,7 @@ class Group extends Value
     public function getItemValue($item)
     {
-        $group = Str::after($this->name(), 'groups/');
+        $group = Str::after($this->name(), 'groups->');
```
Revert, as per other comment.

ryanmitchell replied:
Done

jasonvarga commented on src/Stache/Query/UserQueryBuilder.php:
```diff
@@ -32,6 +32,16 @@ protected function getKeysWithWhere($where)
             ->store('users')
             ->index($where['column'])->items();
 
+        if (in_array($where['column'], ['groups', 'roles'])) {
+            $items = $items->map(function ($item) {
+                return $item->map->handle()->values()->toArray();
+            });
+
+            if (is_array($where['values'])) {
+                $where['type'] = 'JsonContains';
+            }
+        }
+
```
I might be missing the point of what this is doing, but I think rather than adding complexity to the query builder just yet, we should just adjust the query within the `users` tag here:

https://github.com/statamic/cms/blob/f6b573ee5046b55f75bb450c69fe2ec5529eca34/src/Tags/Users.php#L22-L28

Instead of querying `->where('group', $group)` (which is obviously wrong - it must be outdated code), we could instead use `->where('groups/groupname', true)`

Since you added `orWhere` support, we could loop over all the provided groups in the parameter.

```php
foreach ($this->params->explode('group', []) as $i => $group) {
    $method = $i === 0 ? 'where' : 'orWhere';
    $query->$method('groups/'.$group, true);
}
```

(That is untested - just wrote it on the fly in this comment)

ryanmitchell replied:
Ok I've updated the tag to work this way (we dont need to worry about the method approach as orWhere will not care).

jasonvarga commented on src/Stache/Stores/UsersStore.php:
```diff
@@ -12,7 +12,7 @@ class UsersStore extends BasicStore
     {
         return UserGroup::all()->mapWithKeys(function ($group) {
-            return ['groups/'.$group->handle() => Group::class];
+            return ['groups->'.$group->handle() => Group::class];
```
Revert, as per other comment.

ryanmitchell replied:
Done

jasonvarga reviewed (changes requested):
Made a few change requests in individual comments. Happy for pushback.

> Not sure how this fits in with Eloquent

This whole group query thing was clearly not very well thought out. 😞 
Probably ideally we'd be able to treat it more like a relationship somehow. Or maybe do a similar thing to entries and the `whereTaxonomy` method. We could get the user ids in the appropriate groups then have a `whereIn('id', all the ids from the groups)`

ryanmitchell commented:
Asides from the individual comment replies....

Bigger picture, while this works, the `->where('groups/{slug}')` approach seems too stache-specific to me, I agree it would be better to try and aim for a `whereIn('groups', [ids])`, `whereHas('groups', ...)` or even a custom `whereGroups()` approach, so that its compatible with eloquent users. At the moment this tag wouldn't work at all with them (?). 

It would need brought into the UserQueryBuilder so that it can be overridden by Eloquent. 

Something similar also would need implemented for roles.

Do you have a preference?

ryanmitchell commented:
@jasonvarga I've gone ahead and added `whereRole`, `whereGroup` etc to the query builder to be used for role and group querying.

This builds on the previous work in that for file based users it still uses the `->where('group/x')` approach, but it also then allows eloquent to use its own approach. You'll see I've added methods there which should cover the common scenarios, but  devs would be free to change these methods if they need something different.

jmalko commented:
Please consider this PR - this documented functionality is broken in today's version of Statamic.

edalzell commented:
Sorry, I retargeted onto 4.x, locally resolved the conflicts (perhaps incorrectly), but I'm getting an error now, `Call to undefined method App\Models\User::roles()`.

<img width="1746" alt="CleanShot 2023-07-05 at 15 02 47@2x" src="https://github.com/statamic/cms/assets/6069653/ef91e358-155e-4b2b-9e27-edad5226a464">

Maybe the model is missing a relationship?

jasonvarga commented:
@edalzell I resolved the merge and pushed up the change. I didn't verify that the PR works yet, though. Feel free to try it out now. Maybe your issue was an incorrect merge. This PR is over a year old (sorry 😬) so it might just need more tweaking.

edalzell commented:
When I try a role filter (users in the DB), I get this error, with this patch:

```
[2023-07-06 10:08:06] local.ERROR: Call to undefined method App\Models\User::roles() {"userId":"cba0aed3-26f9-4ca9-a67f-8261a605d37b","exception":"[object] (BadMethodCallException(code: 0): Call to undefined method App\\Models\\User::roles() at /Users/erin/Sites/zakat/vendor/laravel/framework/src/Illuminate/Support/Traits/ForwardsCalls.php:67)
[stacktrace]
#0 /Users/erin/Sites/zakat/vendor/laravel/framework/src/Illuminate/Support/Traits/ForwardsCalls.php(36): Illuminate\\Database\\Eloquent\\Model::throwBadMethodCallException('roles')
#1 /Users/erin/Sites/zakat/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(2335): Illuminate\\Database\\Eloquent\\Model->forwardCallTo(Object(Illuminate\\Database\\Eloquent\\Builder), 'roles', Array)
#2 /Users/erin/Sites/zakat/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Concerns/QueriesRelationships.php(868): Illuminate\\Database\\Eloquent\\Model->__call('roles', Array)
#3 /Users/erin/Sites/zakat/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Relations/Relation.php(110): Illuminate\\Database\\Eloquent\\Builder->Illuminate\\Database\\Eloquent\\Concerns\\{closure}()
#4 /Users/erin/Sites/zakat/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Concerns/QueriesRelationships.php(867): Illuminate\\Database\\Eloquent\\Relations\\Relation::noConstraints(Object(Closure))
#5 /Users/erin/Sites/zakat/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Concerns/QueriesRelationships.php(39): Illuminate\\Database\\Eloquent\\Builder->getRelationWithoutConstraints('roles')
#6 /Users/erin/Sites/zakat/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Concerns/QueriesRelationships.php(152): Illuminate\\Database\\Eloquent\\Builder->has('roles', '>=', 1, 'and', Object(Closure))
#7 /Users/erin/Sites/zakat/vendor/statamic/cms/src/Query/EloquentQueryBuilder.php(41): Illuminate\\Database\\Eloquent\\Builder->whereHas('roles', Object(Closure))
#8 /Users/erin/Sites/zakat/vendor/statamic/cms/src/Auth/Eloquent/UserQueryBuilder.php(40): Statamic\\Query\\EloquentQueryBuilder->__call('whereHas', Array)
#9 /Users/erin/Sites/zakat/vendor/statamic/cms/src/Query/Scopes/Filters/UserRole.php(33): Statamic\\Auth\\Eloquent\\UserQueryBuilder->whereRole('field_rep')
```

ryanmitchell commented:
Yes this assumes you define a roles and groups relation on your user model. That would need documented.

edalzell commented:
> Yes this assumes you define a roles and groups relation on your user model. That would need documented.

But how would that work? My roles and groups are flat file, not in the DB. Only my users are in the DB.

ryanmitchell commented:
Hmm my memory is hazy on this and I’m not at my machine to dive into the code. I’ll take a fresh look once I’m back.

ryanmitchell commented:
Ok @edalzell @jasonvarga I've updated this to not require a relation to be set. You'll see in the eloquent UserQueryBuilder it now checks if a relation exists, and if not it performs a whereExists on the role_user or group_user table.

This should cover off the default usage of people switching over from stache. If a dev wants to use a different approach for roles/groups (eg [my PR](https://github.com/statamic/cms/pull/5686)) they would need to define the relationship on their user model.

GMAstengo commented on src/Auth/Eloquent/UserQueryBuilder.php:
```diff
@@ -2,25 +2,153 @@
+        $this->$method(function ($query) use ($operator, $value) {
+            $query->select(DB::raw(1))
+                ->from($this->groupsTable())
+                ->where('groups_id', $operator, $value)
```
I think this should be `group_id`?

ryanmitchell replied:
Yep thanks

GMAstengo commented on src/Auth/Eloquent/UserQueryBuilder.php:
```diff
@@ -2,25 +2,153 @@
+        $this->$method(function ($query) use ($groups) {
+            $query->select(DB::raw(1))
+                ->from($this->groupsTable())
+                ->whereIn('groups_id', $groups)
```
I think this should be `group_id`?

jasperjorna commented:
This one is quite breaking at the moment when using Eloquent for users. Can we expect this to be included in a hotfix release?

jasonvarga commented:
It won't be a hotfix, it'll be a minor release.

edalzell commented:
> This one is quite breaking at the moment when using Eloquent for users. Can we expect this to be included in a hotfix release?

In the meantime you can composer patch it in to start using it.

eiiot commented:
Anyone know when this will be released?

edalzell commented:
It was released a while back

duncanmcclean commented:
> Anyone know when this will be released?

This was released as part of [v4.46.0](https://github.com/statamic/cms/releases/tag/v4.46.0).
